File: Codes/Slurpy/Slurpy.hs
import Text.Regex
import Text.Regex.Posix
import Control.Monad

-- REGULAR EXPRESSIONS
reSlump = makeRegex "([D|E]F+)+G"::Regex
reSlimp = makeRegex "AH|A([D|E]F+)+GC|AB(.+)C"::Regex

-- check functions
isMatched (Just (a,b,c,d)) = (a == "" && b /= "")
isMatched Nothing = False

isFullMatched (Just (a,b,c,d)) = (a == "" && b /= "" && c == "")
isFullMatched Nothing = False

getRecursiveSlimp (Just (a,b,c,d)) = if length d > 1 then d !! 1 else ""
getRecursiveSlimp Nothing = ""

restTestCase (Just (a,b,c,d)) = c
restTestCase Nothing = ""

-- test Slump, Slimp
test regex testCase = if matchedGroup /= "" then test regex matchedGroup else isMatched result
    where
        result = matchRegexAll regex testCase
        matchedGroup = getRecursiveSlimp result

-- test Slurpy
testSlurpy testCase = if isSlimp then isSlump else False
    where
        resultSlimp = matchRegexAll reSlimp testCase
        isSlimp = isMatched resultSlimp
        isSlump = isFullMatched $ matchRegexAll reSlump (restTestCase resultSlimp)


main = do
    -- test case
    let 
        testCaseSlumps = ["DFG", "EFG", "DFFFFFG", "DFDFDFDFG", "DFEFFFFFG"]
        testCaseNotSlumps = ["DFEFF", "EFAHG", "DEFG", "DG", "EFFFFDG"]
        testCaseSlimps = ["AH", "ABAHC", "ABABAHCC", "ADFGC", "ADFFFFGC", "ABAEFGCC", "ADFDFGC"]
        testCaseNotSlimps = ["ABC", "ABAH", "DFGC", "ABABAHC", "SLIMP", "ADGC"]
        testCaseSlurpys = ["AHDFG", "ADFGCDFFFFFG", "ABAEFGCCDFEFFFFFG"]
        testCaseNotSlurpys = ["AHDFGA", "DFGAH", "ABABCC"]
    putStrLn "-----------------------------------------------------"
    putStrLn $ "testCaseSlumps : " ++ show (all (test reSlump) testCaseSlumps) ++ " " ++ show testCaseSlumps
    putStrLn $ "testCaseNotSlumps : " ++ (show . not) (all (test reSlump) testCaseNotSlumps) ++ " " ++ show testCaseNotSlumps
    putStrLn $ "testCaseSlimps : " ++ show (all (test reSlimp) testCaseSlimps) ++ " " ++ show testCaseSlimps
    putStrLn $ "testCaseNotSlimps : " ++ (show . not) (all (test reSlimp) testCaseNotSlimps) ++ " " ++ show testCaseNotSlimps
    putStrLn $ "testCaseSlurpys : " ++ show (all testSlurpy testCaseSlurpys) ++ " " ++ show testCaseSlurpys
    putStrLn $ "testCaseNotSlurpys : " ++ (show . not) (all testSlurpy testCaseNotSlurpys) ++ " " ++ show testCaseNotSlurpys
    
    -- main program
    putStrLn "-----------------------------------------------------"
    putStr "input test count : "
    n <- getLine
    testCases <- forM [1..(read n::Int)] $ \i -> getLine
    -- result
    putStrLn "-----------------------------------------------------"
    putStrLn "SLURPYS OUTPUT"
    mapM_ putStrLn $ map (\testCase -> if (testSlurpy testCase) then "YES" else "NO") testCases
    putStrLn "END OF OUTPUT"

File: Codes/Slurpy/Slurpy.py
import regex as re

# regular expression
reSlump = re.compile("([D|E]F+G*)+G")
reSlimp =  re.compile("AH|A([D|E]F+G*)+GC|AB(?R)C")

# test slurpy
def testSlurpy(testCase):
    m = re.match(reSlimp, testCase)
    if m:
        postFix = testCase[len(m.group()):]
        return re.fullmatch(reSlump, postFix) != None
    else:
        return False

# TestCase
testSlumps = "DFG", "EFG", "DFFFFFG", "DFDFDFDFG", "DFEFFFFFG"
testNotSlumps = "DFEFF", "EFAHG", "DEFG", "DG", "EFFFFDG"
testSlimps = "AH", "ABAHC", "ABABAHCC", "ADFGC", "ADFFFFGC", "ABAEFGCC", "ADFDFGC"
testNotSlimps = "ABC", "ABAH", "DFGC", "ABABAHC", "SLIMP", "ADGC"
testSlurpys = "AHDFG", "ADFGCDFFFFFG", "ABAEFGCCDFEFFFFFG"
testNotSlurpys = "AHDFGA", "DFGAH", "ABABCC"
print("testSlumps :", all([re.fullmatch(reSlump, testCase) != None for testCase in testSlumps]), testSlumps)
print("testNotSlumps :", all([re.fullmatch(reSlump, testCase) == None for testCase in testNotSlumps]), testNotSlumps)
print("testSlimps :", all([re.fullmatch(reSlimp, testCase) != None for testCase in testSlimps]), testSlimps)
print("testNotSlimps :", all([re.fullmatch(reSlimp, testCase) == None for testCase in testNotSlimps]), testNotSlimps)
print("testSlurpys :", all([testSlurpy(testCase) for testCase in testSlurpys]), testSlurpys)
print("testNotSlurpys :", all([not testSlurpy(testCase) for testCase in testNotSlurpys]), testNotSlurpys)
print("-"*50)

if __name__ == "__main__":
    n = input("input test count : ")
    testCases = [input("Test case " + str(i+1) + " : ").upper() for i in range(int(n))]
    print("-"*50)
    print("SLURPYS OUTPUT")
    for testCase in testCases:
        print("YES" if testSlurpy(testCase) else "NO")
    print("END OF OUTPUT")